src/hooks.tsx

```diff
@@ -65,3 +75,136 @@ export function hasScrolled(y = 0): boolean {
+    }
+  }
+
+  return handles;
+}
+
+export function useHashChange(): string {
+  const [hash, setHash] = useState('');
+
+  function onHashChange() {
+    // TODO: probably have to handle deeper index files, and redirects somewhere
+    const newHash =
+      window.location.hash.replace('#', '') === '/'
+        ? '/index'
+        : window.location.hash.replace('#', '');
+    return setHash(newHash);
+  }
+
+  useEffect(() => {
+    window.addEventListener('hashchange', onHashChange);
+    return () => window.removeEventListener('hashchange', onHashChange);
+  }, []);
+
+  return hash;
+}
+
+export function useDirectorySelector(): {
+  select: () => void;
+  handles: FileSystemFileHandles;
+  error: Error;
+  pending: boolean;
+  config: ProjectConfig;
+} {
+  const [error, setError] = useState<Error | null>(null);
+  const [pending, setPending] = useState(false);
+  const [handles, setHandles] = useState<FileSystemFileHandles | null>();
+  const [config, setConfig] = useState<ProjectConfig | null>();
+
+  const select = useCallback(async () => {
+    setPending(true);
+    try {
+      const handle = await window.showDirectoryPicker();
+      let docs: FileSystemDirectoryHandle;
+      let foundDocsJson = false;
+      for await (const entry of handle.values()) {
+        if (entry.kind === 'file' && entry.name === 'docs.json') {
+          const configFile = await entry.getFile();
+          const configText = JSON.parse(await configFile.text());
+
+          setConfig(mergeConfig(configText));
+          foundDocsJson = true;
+        }
+        if (entry.kind === 'directory' && entry.name === 'docs') {
+          docs = entry;
+        }
+      }
+
+      if (!foundDocsJson) {
+        throw new Error('No docs.json found');
+      }
+
+      if (!docs) {
+        throw new Error('No docs directory found');
+      }
+
+      // TODO set config
+      setHandles(await iterateDirectory(docs));
+    } catch (e) {
+      setError(e);
+    } finally {
+      setPending(false);
+    }
+  }, []);
+
+  return { select, handles, error, pending, config };
+}
+
+export function usePollLocalDocs(
+  handles: FileSystemFileHandles,
+  config: ProjectConfig,
+  ms = 500,
+  setPageProps: React.Dispatch<React.SetStateAction<PreviewPageProps>>,
+): void {
+  const hash = useHashChange();
+
+  useEffect(() => {
+    if (!handles) {
+      return;
+    }
+    const handle = hash !== '' ? handles[hash] : handles['/index'];
+
+    // TODO handle no file (404)
+    if (!handle) {
+      console.log('file not found, 404');
+      return;
+    }
+
+    // TODO update state and show page
+    const interval = setInterval(
+      () =>
+        handle
+          .getFile()
+          .then(file => file && file.text())
+          .then(text => text && buildPreviewProps({ hash, config: JSON.stringify(config), text }))
+          .then(props => props && setPageProps(props)),
+      ms,
+    );
+
+    return () => clearInterval(interval);
+  }, [handles, hash]);
```

Comment:
Diffing on `handles` wont work here, since it's an object. I think it could probably get away with only doing this on a hash change - if the hash changes thats the only time we want to kill the polling right?